### PR TITLE
Calibrate FedEx

### PR DESCRIPTION
**Pull request by joneswong**
In the original implementation of FedEx, the policy is allowed to use after - before as reward, where after and before refer to the validation loss estimated before after and before the local updates.

This pr extends the base torch trainer so that we can optionally conduct evaluation before and after the local updates.

### CODE REVIEW

**Comment by joneswong** [comment]
@DavdGao @rayrayraykk As the torch trainer is moved to a new .py file in this pr, all the code is regarded as changes. Please focus on the `train()` method, where the actual changes happen.

**Review by DavdGao — COMMENTED** [review]
Looks good to me

**Review by rayrayraykk — APPROVED** [review]
Looks good to me.
